lib/ansible/modules/network/nxos/nxos_rpm.py

```diff
@@ -0,0 +1,271 @@
+                time.sleep(2)
+                iteration += 1
+            else:
+                return
+        else:
+            return
+
+
+def validate_operation(module, show_cmd, cfg_cmd, pkg, pkg_not_present):
+    iteration = 0
+    while iteration < 10:
+        body = execute_show_command(show_cmd, module)
+        if pkg_not_present:
+            if pkg not in body:
+                return
+        else:
+            if pkg in body:
+                return
+        time.sleep(2)
+        iteration += 1
+
+    err = 'Operation "{0}" Failed'.format(cfg_cmd)
+    module.fail_json(msg=err)
+
+
+def add_operation(module, show_cmd, file_system, full_pkg, pkg):
+    cmd = 'install add {0}:{1}'.format(file_system, full_pkg)
+    config_cmd_operation(module, cmd)
+    validate_operation(module, show_cmd, cmd, pkg, False)
+    return cmd
+
+
+def activate_operation(module, show_cmd, pkg):
+    cmd = 'install activate {0} forced'.format(pkg)
+    config_cmd_operation(module, cmd)
+    validate_operation(module, show_cmd, cmd, pkg, False)
+    return cmd
+
+
+def commit_operation(module, show_cmd, pkg):
+    cmd = 'install commit {0}'.format(pkg)
+    config_cmd_operation(module, cmd)
+    validate_operation(module, show_cmd, cmd, pkg, False)
+    return cmd
+
+
+def deactivate_operation(module, show_cmd, pkg, flag):
+    cmd = 'install deactivate {0} forced'.format(pkg)
+    config_cmd_operation(module, cmd)
+    validate_operation(module, show_cmd, cmd, pkg, flag)
+    return cmd
+
+
+def remove_operation(module, show_cmd, pkg):
+    cmd = 'install remove {0} forced'.format(pkg)
+    config_cmd_operation(module, cmd)
+    validate_operation(module, show_cmd, cmd, pkg, True)
+    return cmd
+
+
+def install_remove_rpm(module, full_pkg, file_system, state):
+    commands = []
+
+    splitted_pkg = full_pkg.split('.')
+    pkg = '.'.join(splitted_pkg[0:-1])
+
+    show_inactive = 'show install inactive'
+    show_active = 'show install active'
+    show_commit = 'show install committed'
+    show_patches = 'show install patches'
+
+    if state == 'present':
+        inactive_body = execute_show_command(show_inactive, module)
+        active_body = execute_show_command(show_active, module)
+
+        if pkg not in inactive_body and pkg not in active_body:
+            commands.append(add_operation(module, show_inactive, file_system, full_pkg, pkg))
+
+        if pkg not in active_body:
+            commands.append(activate_operation(module, show_active, pkg))
+
+        commit_body = execute_show_command(show_commit, module)
+        if pkg not in commit_body:
+            patch_body = execute_show_command(show_patches, module)
+            if pkg in patch_body:
+                # this is an smu
+                commands.append(commit_operation(module, show_active, pkg))
+            else:
+                err = 'Operation "install activate {0} forced" Failed'.format(pkg)
+                module.fail_json(msg=err)
```

Comment:
Shouldn't this be caught in `activate_operation`? Am I missing any corner case?

Reply:
* For patch/SMU, there are 3 operations. 
add- which adds the patch and puts it in inactive state
activate - which activates the patch and puts it in active state
commit - which commits the patch and puts it in commit state
* For feature RPMs, there are only 2 operations.
add - which adds the RPM and puts it in inactive state
activate - which activates & commits the RPM and puts it in active and commit state. 

Because the code is common, we are checking after the activate operation, if the package is in active state or not. For SMU, we commit and then check if it is in commit state or not. Since there is no such thing for feature RPM, the expectation is that the platform has already put it commit state also, which is generally the case almost always. However, if there is an issue on the platform, and it put it only in active state but not in commit state, there are no more operations or commands to push it into that commit state. So all we can do is FAIL the operation. If we do not do this, it will be seen as success, which would be incorrect. This is defensive code from our side to take care of an extreme case on the platform.

Hope it is clear now.